lite2/client.go

```diff
@@ -561,15 +561,32 @@ func (c *Client) verifyHeader(newHeader *types.SignedHeader, newVals *types.Vali
 			panic(fmt.Sprintf("Unknown verification mode: %b", c.verificationMode))
 		}
 	} else {
-		// 2) Otherwise, perform backwards verification
-		// Find the closest trusted header after newHeader.Height
-		var closestHeader *types.SignedHeader
-		closestHeader, err = c.trustedStore.SignedHeaderAfter(newHeader.Height)
+		var firstHeaderHeight int64
+		firstHeaderHeight, err = c.FirstTrustedHeight()
 		if err != nil {
-			return errors.Wrapf(err, "can't get signed header after height %d", newHeader.Height)
+			return err
+		}
+		var closestHeader *types.SignedHeader
+		if newHeader.Height < firstHeaderHeight {
+			c.logger.Info("Using backwards")
+			closestHeader, err = c.TrustedHeader(firstHeaderHeight)
+			if err != nil {
+				return errors.Wrapf(err, "can't get signed header after height %d", newHeader.Height)
+			}
+			err = c.backwards(closestHeader, newHeader, now)
+			//if err != nil { return err }
+		} else {
+			closestHeader, err = c.trustedStore.SignedHeaderBefore(newHeader.Height)
+			if err != nil {
+				return errors.Wrapf(err, "can't get signed header before height %d", newHeader.Height)
+			}
+			var closestValidatorSet *types.ValidatorSet
+			closestValidatorSet, _, err = c.TrustedValidatorSet(closestHeader.Height)
+			if err != nil {
+				return errors.Wrapf(err, "can't get validator set at height %d", closestHeader.Height)
+			}
+			err = c.bisection(closestHeader, closestValidatorSet, newHeader, newVals, now)
```

Comment:
should we fallback to backwards verification if this fails?

Reply:
should we actually check if closestHeader can be trusted before attempting `bisection`?

Reply:
if it's in the trustedStore then we don't we assumed that is is trusted?

Reply:
if bisection fails it goes to sequential verification automatically which is much the same algorithm as backwards (although backwards is considerably faster)

Reply:
not if we're doing bisection? zarko: "That’s fine if trusted header with bisection is within trusted period."

Reply:
true, I will change that